docs/proposals/20210310-opt-in-autoscaling-from-zero.md

```diff
@@ -0,0 +1,242 @@
+
+## Proposal
+
+To facilitate scaling from zero replicas, the minimal information needed by the cluster autoscaler
+is the CPU and memory resources for the target node group that will be scaled. The autoscaler uses
+this information to create a prediction about how many nodes should be created when scaling. In
+most situations this information can be directly read from the nodes that are running within a
+node group. But, during a scale from zero situation (ie when a node group has zero replicas) the
+autoscaler needs to acquire this information from the cloud provider.
+
+An optional field is proposed for users to add to their Infrastructure Machine Templates which will
+carry information about the CPU, memory, and GPU requirements for the machines created from the template.
+The cluster autoscaler will then utilize this information by reading the appropriate infrastructure
+reference from the resource it is scaling (MachineSet or MachineDeployment).
+
+A user may override the field in the associated infrastructure template by applying annotations to the
+MachineSet or MachineDeployment in question. In these cases the autoscaler will evaluate the annotation
+information in favor of reading the information from the status.
+
+### User Stories
+
+#### Story 1
+
+As a Cluster API user, I would like to reduce my operating costs by scaling down my workload
+cluster when they are not in use. Using the cluster autoscaler with a minimum size of zero for
+a MachineSet or MachineDeployment will allow me to automate the scale down actions for my clusters.
+
+#### Story 2
+
+As a Cluster API user, I would like to have special resource nodes (eg GPU enabled) provided when needed by workloads
+without the need for human intervention. As these nodes might be more expensive, I would also like to return them when
+not in use. By using the cluster autoscaler with a zero-sized MachineSet or MachineDeployment, I can automate the
+creation of nodes that will not consume resources until they are required by applications on my cluster.
+
+### Implementation Details/Notes/Constraints
+
+Cloud providers, or users, can add an optional field to their machine template infrastructure objects to indicate
+the resources provided by the machine. Internally, this field will be represented by a Go `map` type
+utilizing named constants for the keys and `k8s.io/apimachinery/pkg/api/resource.Quantity` as the values
+(similar to how resource limits and requests are handled for pods).
+
+**Example implementation in Docker provider**
+```
+// these constants will be carried in Cluster API, but are repeated here for clarity
+const (
+    AutoscalerResourceCPU ResourceName = "cpu"
+    AutoscalerResourceMemory ResourceName = "memory"
+    AutoscalerResourceGPU ResourceName = "gpu"
+)
+
+// DockerMachineTemplateSpec defines the desired state of DockerMachineTemplate.
+type DockerMachineTemplateSpec struct {
+    AutoscalerCapacity ResourceList `json:"autoscalerCapacity,omitempty"`
```

Comment:
I know that it introduces some nesting, some struct nesting, but should this be `spec.autoscaler.machine.request`...

Reply:
i had thought about something like this, but wasn't sure how much nesting we wanted XD

i will sketch up a version that uses this structure so we can see how it looks.

Reply:
This information is only used when scaling from zero right? I wonder if there's a way to make this more explicit in the field names that it's only used in that situation (and e.g. not when scaling further up)

Reply:
yeah, it's a tough problem from a naming standpoint. i had originally thought using the word "hint" might be helpful, but after looking through the kubernetes docs, i feel like "capacity" is probably closer to what we are asking for here.